hmmlearn/hmm.py

```diff
@@ -520,134 +520,221 @@ class GMMHMM(_BaseHMM):
-        return self.gmms_[state].sample(1, random_state=random_state).flatten()
+        # TODO: actually use random_state
+        if random_state is None:
+            random_state = self.random_state
+
+        cur_means = self.means_[state]
+        cur_covs = self.covars_[state]
+        cur_weights = self.weights_[state]
+
+        i_gauss = np.random.multinomial(1, cur_weights)[0][0]
+        mean = cur_means[i_gauss]
+        if self.covariance_type == 'tied':
+            cov = cur_covs
+        else:
+            cov = cur_covs[i_gauss]
+
+        return sample_gaussian(mean, cov, self.covariance_type,
+                               random_state=random_state)
+
+    def _compute_log_likelihood(self, X):
+        n_samples, _ = X.shape
+        res = np.zeros((n_samples, self.n_components))
+
+        for i in range(self.n_components):
+            cur_means = self.means_[i]
+            cur_covs = self.covars_[i]
+            if self.covariance_type == 'spherical':
+                cur_covs = cur_covs[:, np.newaxis]
+            cur_weights = self.weights_[i]
+
+            densities = np.exp(log_multivariate_normal_density(
+                X, cur_means, cur_covs, self.covariance_type
+            ))
+            res[:, i] = np.log(np.sum(cur_weights * densities, axis=1))
+
+        return res
 
     def _initialize_sufficient_statistics(self):
         stats = super(GMMHMM, self)._initialize_sufficient_statistics()
-        stats['norm'] = [np.zeros(g.weights_.shape) for g in self.gmms_]
-        stats['means'] = [np.zeros(np.shape(g.means_)) for g in self.gmms_]
-        stats['covars'] = [np.zeros(np.shape(g.covars_)) for g in self.gmms_]
+        stats['n_samples'] = 0
+        stats['post_comp_mix'] = None
+        stats['post_mix_sum'] = np.zeros((self.n_components, self.n_mix))
+        stats['post_sum'] = np.zeros(self.n_components)
+        stats['samples'] = None
+        stats['centered'] = None
         return stats
 
     def _accumulate_sufficient_statistics(self, stats, X, framelogprob,
-                                          posteriors, fwdlattice, bwdlattice):
+                                          post_comp, fwdlattice, bwdlattice):
+
+        # TODO: support multiple frames
+
         super(GMMHMM, self)._accumulate_sufficient_statistics(
-            stats, X, framelogprob, posteriors, fwdlattice, bwdlattice)
+            stats, X, framelogprob, post_comp, fwdlattice, bwdlattice
+        )
+
+        n_samples, _ = X.shape
+
+        stats['n_samples'] = n_samples
+        stats['samples'] = X
+
+        # post_comp = posteriors.reshape((n_samples,))
+
+        eps = 1e-15
+
+        prob_mix = np.zeros((n_samples, self.n_components, self.n_mix))
+        for p in range(self.n_components):
+            covar = self.covars_[p]
+            if self.covariance_type == 'spherical':
+                covar = covar[:, np.newaxis]
+            prob_mix[:, p, :] = np.exp(log_multivariate_normal_density(
```

Comment:
Maybe we should have the mixture density as a separate function since it's computed in both `_compute_log_likelihood` and `_accumulate_sufficient_statistics`.